### DISCUSSION TITLE
Consider making use of kotlin-power-assert which will become part of the standard compiler plugin set (KT-63607)

### DISCUSSION THREAD

**Discussion by LeoColman** [kotest/kotest#3952] [6.0 changes]
[REF](https://github.com/kotest/kotest/issues/3918#issuecomment-1987227183)

Consider making use of kotlin-power-assert which will become part of the standard compiler plugin set ([KT-63607](https://youtrack.jetbrains.com/issue/KT-63607/Migrate-kotlin-power-assert-into-Kotlin-repository)).

**Comment by LeoColman** [comment]
[REF](https://github.com/kotest/kotest/issues/3918#issuecomment-1987320495)

> * Consider making use of kotlin-power-assert which will become part of the standard compiler plugin set ([KT-63607](https://youtrack.jetbrains.com/issue/KT-63607/Migrate-kotlin-power-assert-into-Kotlin-repository)).

Should we ask them to consider supporting other types of function names, so we could make it work with shouldBe ?

**Comment by LeoColman** [comment]
[REF](https://github.com/kotest/kotest/issues/3918#issuecomment-1987349442)

Seems like functionality like this is already there: https://github.com/bnorm/kotlin-power-assert
> The Gradle plugin allows configuring the functions which should be transformed with a list of fully-qualified function names.
```kotlin
// Kotlin DSL
configure<com.bnorm.power.PowerAssertGradleExtension> {
  functions = listOf("kotlin.assert", "kotlin.test.assertTrue")
}
```

**Reply by LeoColman** [reply]
[REF](https://github.com/kotest/kotest/issues/3918#issuecomment-1987366851)

> Seems like functionality like this is already there: https://github.com/bnorm/kotlin-power-assert
> 
> > The Gradle plugin allows configuring the functions which should be transformed with a list of fully-qualified function names.
> 
> ```kotlin
> // Kotlin DSL
> configure<com.bnorm.power.PowerAssertGradleExtension> {
>   functions = listOf("kotlin.assert", "kotlin.test.assertTrue")
> }
> ```

I wounder if it will work for infix though?

**Reply by twadzins** [reply]
From what I see in the [newly official docs](https://kotlinlang.org/docs/power-assert.html#beyond-assert-function), it requires "taking a String or () -> String value as the last parameter", which might be tricky/infeasible to make work with infix methods. 

>Functions like require(), check(), assertTrue(), assertEqual() and others can also be transformed, if they have a form that allows taking a String or () -> String value as the last parameter.

I'm hopeful that this limitation could be removed if there isn't a way to make it work as is, since having the power assert error message feature and still being able to use kotest assertions would be amazing.

Edit:
I do see infix functionality seemingly working on the original kotlin-power-assert repo:
https://github.com/bnorm/kotlin-power-assert/issues/99 

That link points to a fixed issue on the official plugin:
https://youtrack.jetbrains.com/issue/KT-65640/PowerAssert-Infix-function-not-aligned-correctly

I'll try playing with this later when I get some time.

**Comment by sksamuel** [comment]
We could fork the plugin.

**Comment by sksamuel** [comment]
We can make it work out of the box 🎉
https://github.com/kotest/kotest-power-assert

**Reply by CLOVIS-AI** [reply]
Is there a reason in particular you're printing the message and not using `withClue`? Since this is just printing, I'm not sure it's going to work well with soft assertions etc.

**Reply by sksamuel** [reply]
I was just testing out that it would work.

**Reply by bnorm** [reply]
Something you can consider, if you want to reduce the impact on your public API, is mark the power-assert overloads as `internal` and/or `@Deprecated`:

```kotlin
@Deprecated(message = "Used by power-assert", level = DeprecationLevel.HIDDEN)
@Suppress("unused")
internal fun <T> T.shouldBe(expected: T, msg: String) {
```

**Reply by sksamuel** [reply]
Good idea, thanks @bnorm

**Comment by sksamuel** [comment]
![image](https://github.com/kotest/kotest/assets/743706/7d657c16-a0aa-490b-909f-19cb322c1e3d)

**Comment by sksamuel** [comment]
Closing as power assert is supported by kotest since 6.0
